### ISSUE TITLE
JSON format

### ISSUE THREAD

**Issue by jetuk** [pywr/pywr#123]
Re `json`  branch we need to agree a (minimum) feature set to implement before this is merged.

- [x] Network structure of nodes/links
    - [x] Edge slots
- [x] Model parameters (keywords)
- [ ] Node `Parameter` objects
    - [x] External timeseries support
        - [x] Excel
        - [x] CSV
        - [x] HDF5
        - [x] gzipped csv
    - [x] Monthly profiles
    - [ ] Reservoir control curves
- [x] `Recorder` objects
    - [x] NumpyArray*Recorder
    - [x] Aggregated recorders
    - [x] Other recorders
- [x] Node positions
- [ ] Domains
- [ ] Scenarios
- [x] Load the above
- [ ] Save the above

Also:

- [x] Port existing XML models to JSON

Some questions to that we need to resolve:
- How to deal with `Parameter` and `Recorder` subclasses? Do they need class specific implementations?
- How to deal with customisation / scripting of a model. Should any of this be saved? Perhaps not in the immediate term.

**Comment by snorfalorpagus** [comment]
I don't think –at least in the short term– that there should be any Python stored in the JSON file. If there was a UI to manage this it might be different. But until then I think it should live alongside the JSON file in a separate .py file.

From the work I've done so far I think saving of JSON might be much harder to implement than loading.

Not sure what you mean about the Parameter/Recorder subclasses.

**Comment by jetuk** [comment]
I agree. 

Re  Parameter/Recorder subclasses I only meant that some of them have bespoke `__init__` arguments. So how would the loader initialise them without a per class implementation?

**Comment by snorfalorpagus** [comment]
One approach is to pass any attributes as keyword arguments.

e.g. `<parameter name="somename" />` becomes `parameter = Parameter(name="somename")`.

I can't see how this can work for saving though without class specific implementations.

**Comment by snorfalorpagus** [comment]
I've pushed a couple of more changes to the `json` branch. I think it's 50% done. Hopefully finish this in the next couple of weeks.

**Comment by snorfalorpagus** [comment]
I'm thinking about how we represent scenarios in JSON. How does this look:

```
"scenarios": {
    "inflow": { "size": 2 },
},
"nodes": {
    "catchment": {
        "max_flow": {
            "type": "scenarioconstant",
            "scenario": "inflow":
            "values": [5.0, 10.0],
        }
    },
    "demand": {
        "max_flow": 10,
        "cost": -999,
    }
}
```

Probably need something more complicated for the more complicated scenario parameters (e.g. `ArrayIndexedScenarioMonthlyFactorsParameter`).

**Comment by jetuk** [comment]
I assume for the more complex parameters we need a "kwargs" type item to pass additional information to the constructor? I guess those could get very complex?

Otherwise it looks fine. The only other option would be to have a "parameters" section which defined a whole load of parameters (by name or id or something?) that then get mapped to individual nodes. For example how would this deal with sharing the same `Parameter` between two nodes or attributes?

**Comment by snorfalorpagus** [comment]
I think we can probably support a mixture of inline parameters (the vast majority) and references to parameters in a list somewhere. The parameters with references will need a unique name. This gets more complicated when you consider *writing* a JSON file. As a general rule, can parameters be shared between nodes?

```
"nodes": {
    "somenode": {
        "max_flow": "refname",
        "cost": 42,
    }
},
"parameters": {
    "refname": {
        ....
    }
}
```

**Comment by snorfalorpagus** [comment]
Added support for parameters by reference (as the comment above) in ac46e63.

**Comment by snorfalorpagus** [comment]
I think the next major step is to move the code from `load_parameter` into class methods of `Parameter` and it's subclasses. In order to do this I think we need some sort of "parameter registry" similar to the one we have for solvers, so that we can identify them using a string from JSON. Cython classes can't use metaclasses other than type, so I think we'll need something like this:

```
cdef ArrayIndexedParameter(Parameter):
    ...
parameter_registry.add(ArrayIndexedParameter)
```

i.e. manually registering parameters after they are defined. This is required so that users can create new parameters that are recognised by JSON without having to recompile any part of pywr.

In addition to this we might need better tracking of parameter names. I've done this with a simple dictionary so far, but it's only created when reading a JSON file. This will need to be done better if we're to support writing JSON.

**Comment by jetuk** [comment]
For writing what will need to be better about it?

Do we need the same approach for `Recorder` ?

**Comment by snorfalorpagus** [comment]
> For writing what will need to be better about it?

The parameter names will need to be unique. Currently there is nothing enforcing this.

I've ported the existing XML models to JSON now – for the tests that aren't marked `xfail` at least.

**Comment by khaledk2** [comment]
I have some question about the new json file:
There are 3 parts were not in the previous example which are:
1.	parameters
2.	domains
3.	recorders

I think the parameters part contains values for some nodes attributes, for examples Max-flow for both nodes demand1 and demand2 is equal to demand_max_flow reference value in the parameters sections which is 81 in this case, and so on for the rest parameters which have references in the parameters section, is that correct? 
I am not sure about the other two sections?

In addition, could you please send me link to this file:
"url": "examples/data/thames_stochastic_flow.gz",

Could you please let me know what the data files format which pywr supports?

**Comment by snorfalorpagus** [comment]
You're correct about the parameters section. Parameters can either be defined inline (i.e. within the nodes section) or defined in the parameters section and referenced by name in the node. The need for this arises from re-use of parameters, e.g. for both the max_flow and min_flow properties of a node.

The domains section will be used to describe domains used in the model. A domain is used to separate two networks in the same model that cannot transfer flow directly. For example, a model could represent flow of water and electricity. It's possible to convert flow from one network to another. In this example a hydropower scheme could convert flow of water to electricity.

Recorders keep a record of variables (such as flow, volume or level) as the model runs. They're used to store model results.

In theory we can support any file format that pandas can read. Currently it supports CSV and Excel, with plans for HDF5 also.

The file you asked for is in the main branch, here: https://github.com/pywr/pywr/blob/master/examples/data/thames_stochastic_flow.gz

We have need for the JSON format to be working in a couple of weeks time so it should be merged by then. Then there should be some time to write some proper documentation for all of this!

**Comment by snorfalorpagus** [comment]
I think this is ready for merge into master, with the following improvements still required which I propose we create separate issues for (in rough order of importance):

- Reservoir control curves - we still need to agree on how our default `Reservoir` class behaves. There are some differences between the existing control curve structure and what I've used in the WaterUK project.
- Timeseries - currently no support for data that isn't indexed contiguously 0...N in alignment with the model run. In addition to #127 we'll need some minor changes to `load_parameter_values`.
- Scenarios - I don't think this will be difficult to implement, but I don't have time to do it by Wednesday
- Domains - not a priority as we don't use them currently
- Saving JSON - I still think this is going to be tricky. Some issues including preservation of comments, and preservation of the order of information to produce useful diffs when version controlling models.

There are some tests ready, as the existing XML models have been converted to JSON (unless they were marked xfail in pytest - e.g. river splits). Otherwise it's difficult to test everything without proper coverage reports (see #86).

I'm going to attempt to squash and merge now.

**Comment by snorfalorpagus** [comment]
Merged as ac4610e5bfb66ae1fc4fe44294369d0e5fbf5a2f.

Further discussion of the issues above should be done in new tickets.
